### PR TITLE
[p2p] getblock for 1-block reorgs in response to compact block message

### PR DESCRIPTION
**Pull request by instagibbs**
Our headers announcement logic will download any blocks that match the total work: https://github.com/bitcoin/bitcoin/blob/8b262eb2d80bfa27ae8501078ce47bc1407e9c55/src/net_processing.cpp#L1443

Compact block responses are asymmetric with this: We only intentionally download the block when the new chain has more work, or when it's already marked as "in flight".

A wrinkle here is that the new behavior will only affect high bandwidth compact block peers: currently low bandwidth compact block peers have already marked this particular block as "in flight", which allows the full block to be downloaded with this catch-all "some reason" logic. For high bandwidth compact block peers this message is the first the block has been heard of it, and previous behavior is to simply ignore the compact block announcement.

This change synchronizes the treatment of new block announcements.

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -2406,10 +2406,11 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return true;
 
-        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know something better
+        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know same or better better
```
Comment (practicalswift) [inline src/net_processing.cpp]: Nit: "better better" should be something else? :-)

**Comment by instagibbs** [comment]
@practicalswift fixed

**Comment by rebroad** [comment]
why not simply request the compact block rather than a full block?

**Comment by TheBlueMatt** [comment]
Should mark the block as in flight if it wasn't already and we requested it. Probably should be checking if it's in flight from the same peer too (which would be required if you mark it, and should probably have been done either way).

@promag because the mempool reconstruction will be useless so it'll be worse than a normal request.

**Comment by instagibbs** [comment]
@TheBlueMatt Added the filter-by-sender and marking block as in flight.

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2406,11 +2406,15 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return true;
 
-        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know something better
+        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know same or better
                 pindex->nTx != 0) { // We had this block at some point, but pruned it
-            if (fAlreadyInFlight) {
-                // We requested this block for some reason, but our mempool will probably be useless
+            if ((fAlreadyInFlight && blockInFlightIt->second.first == pfrom->GetId())
```
Comment (Empact) [inline src/net_processing.cpp]: nit: how about unpacking `blockInFlightIt->second.first` to a local like `inFlightNodeId` or the like? It's used one other place in the file and is relatively inscrutable.

Comment (instagibbs) [reply src/net_processing.cpp]: I cached the "from same peer" logic instead since it's used twice.

**Comment by rebroad** [comment]
would it be worth keeping the txs that went into the latest block to hand
for this situation so that compact blocks can be used? what if the previous
compact block received had only one transaction in it, whereas this
competing block has many transactions? by requesting a full block we would
be favouring the empty block when it would be better for the network to
favour the full block. therefore I'd still argue it's better to request a
compact block so to not give empty blocks an unfair advantage.

**Comment by instagibbs** [comment]
@rebroad I think this goes a bit off-topic to incentives. As-is it would likely take a lot of work to support it, and I'd have no idea how to do it safely.

File: src/net_processing.cpp [hunk 3]
```diff
@@ -2402,15 +2402,20 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
...
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return true;
 
-        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know something better
+        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know same or better
                 pindex->nTx != 0) { // We had this block at some point, but pruned it
-            if (fAlreadyInFlight) {
-                // We requested this block for some reason, but our mempool will probably be useless
+            if ((fAlreadyInFlight && in_flight_from_same_peer)
+                    || pindex->nChainWork == chainActive.Tip()->nChainWork) {
```
Comment (instagibbs) [inline src/net_processing.cpp]: I think this will actually request the full block that *is* the known chaintip. Probably need to filter it by blockhash.

Comment (instagibbs) [reply src/net_processing.cpp]: oh, maybe not. two lines or so above we have this check:

if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
            return true;

Comment (TheBlueMatt) [reply src/net_processing.cpp]: I dont think we want to request the block if its already in flight from another peer, even if it has equal work with our tip.

File: src/net_processing.cpp [hunk 4]
```diff
@@ -2402,15 +2402,20 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
...
+                // We requested this block for some reason from the same peer
+                // or just want same height block, but our mempool will probably be useless
                 // so we just grab the block via normal getdata
+                std::list<QueuedBlock>::iterator* queuedBlockIt = nullptr;
```
Comment (TheBlueMatt) [inline src/net_processing.cpp]: This appears to be unused. That said, I think we should consider handling if peer accidentally sends us the block twice, thus queuedBlockIt->partialBlock is set after we call MarkBlockAsInFlight (probably by just wiping the partialBlock).

**Review by TheBlueMatt — COMMENTED** [review]
Indeed, that is what I meant wrt in_flight_from_same_peer. It appears to be indentical to what I did in #10984 at https://github.com/bitcoin/bitcoin/pull/10984/commits/a63474546943ef857554adb68c8c0b23e3a24f88 so it must be right! :p.

File: src/net_processing.cpp (hunk 4 above)
Comment (instagibbs) [reply src/net_processing.cpp]: Ok, replicated the logic from a few lines below covering the same case.

**Comment by instagibbs** [comment]
Comments addressed

File: src/net_processing.cpp [hunk 5]
```diff
@@ -2402,15 +2402,28 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
...
-                // We requested this block for some reason, but our mempool will probably be useless
+            if ((fAlreadyInFlight && in_flight_from_same_peer)
+                    || (!fAlreadyInFlight && pindex->nChainWork == chainActive.Tip()->nChainWork)) {
+                // We requested this block for some reason from the same peer
+                // or just want same height block, but our mempool will probably be useless
                 // so we just grab the block via normal getdata
+                std::list<QueuedBlock>::iterator* queuedBlockIt = nullptr;
+                if (!MarkBlockAsInFlight(pfrom->GetId(), pindex->GetBlockHash(), pindex, &queuedBlockIt)) {
+                    if (!(*queuedBlockIt)->partialBlock) {
+                        (*queuedBlockIt)->partialBlock.reset(new PartiallyDownloadedBlock(&mempool));
```
Comment (TheBlueMatt) [inline src/net_processing.cpp]: You want to reset to nullptr here, as we're about to fetch via non-compact-blocks.

Comment (instagibbs) [reply src/net_processing.cpp]: done

File: src/net_processing.cpp [hunk 6]
```diff
@@ -2402,15 +2402,28 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
 
         std::map<uint256, std::pair<NodeId, std::list<QueuedBlock>::iterator> >::iterator blockInFlightIt = mapBlocksInFlight.find(pindex->GetBlockHash());
         bool fAlreadyInFlight = blockInFlightIt != mapBlocksInFlight.end();
+        bool in_flight_from_same_peer = fAlreadyInFlight ? blockInFlightIt->second.first == pfrom->GetId() : false;
```
Comment (Empact) [inline src/net_processing.cpp]: `&&` over `?` here

File: src/net_processing.cpp [hunk 7]
```diff
@@ -2402,15 +2402,28 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
...
 
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return true;
 
-        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know something better
+        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know same or better
                 pindex->nTx != 0) { // We had this block at some point, but pruned it
-            if (fAlreadyInFlight) {
-                // We requested this block for some reason, but our mempool will probably be useless
+            if ((fAlreadyInFlight && in_flight_from_same_peer)
```
Comment (Empact) [inline src/net_processing.cpp]: Redundant `fAlreadyInFlight` check

File: src/net_processing.cpp [hunk 8]
```diff
@@ -2402,15 +2402,28 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
...
+                    || (!fAlreadyInFlight && pindex->nChainWork == chainActive.Tip()->nChainWork)) {
+                // We requested this block for some reason from the same peer
+                // or just want same height block, but our mempool will probably be useless
                 // so we just grab the block via normal getdata
+                std::list<QueuedBlock>::iterator* queuedBlockIt = nullptr;
```
Comment (Empact) [inline src/net_processing.cpp]: snake_case for new vars

File: src/net_processing.cpp [hunk 9]
```diff
@@ -2432,7 +2445,7 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
         // possibilities in compact block processing...
         if (pindex->nHeight <= chainActive.Height() + 2) {
             if ((!fAlreadyInFlight && nodestate->nBlocksInFlight < MAX_BLOCKS_IN_TRANSIT_PER_PEER) ||
-                 (fAlreadyInFlight && blockInFlightIt->second.first == pfrom->GetId())) {
+                 (fAlreadyInFlight && in_flight_from_same_peer)) {
```
Comment (Empact) [inline src/net_processing.cpp]: Redundant `fAlreadyInFlight` check

**Comment by instagibbs** [comment]
@Empact comments addressed

File: src/net_processing.cpp [hunk 10]
```diff
@@ -2402,15 +2402,28 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
...
 
         if (pindex->nStatus & BLOCK_HAVE_DATA) // Nothing to do here
             return true;
 
-        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know something better
+        if (pindex->nChainWork <= chainActive.Tip()->nChainWork || // We know same or better
                 pindex->nTx != 0) { // We had this block at some point, but pruned it
-            if (fAlreadyInFlight) {
-                // We requested this block for some reason, but our mempool will probably be useless
+            if ((in_flight_from_same_peer)
```
Comment (Empact) [inline src/net_processing.cpp]: nit: unnecessary parens

Comment (instagibbs) [reply src/net_processing.cpp]: done

**Comment by instagibbs** [comment]
I don't think anyone remaining who would review cares about this, closing.

**Pull request closed without merging**